lib/app/endpoint/oauth2_endpoints.rb

```diff
@@ -5,91 +5,74 @@ class App
-                                  Please click your browser's "Back" button to return to Inferno,
-                                  and click "Refresh" to ensure that the most recent test results are visible.
-                                </p>
-                                )
-
-                  error_message += "<p>Error returned by server: <strong>#{params[:error]}</strong>.</p>" if params[:error].present?
-
-                  error_message += "<p>Error description returned by server: <strong>#{params[:error_description]}</strong>.</p>" unless params[:error_description].nil?
-
-                  halt 500, error_message
-                elsif instance&.waiting_on_sequence&.wait?
-                  error_message = "State provided in redirect (#{params[:state]}) does not match expected state (#{instance.state})."
-                else
-                  redirect "#{base_path}/#{cookies[:instance_id_test_set]}/?error=no_state&state=#{params[:state]}"
-                end
-              end
-            end
+          def running_test_found?
+            @instance.present? &&
+              @instance.client_endpoint_key == params[:key] &&
+              %w[launch redirect].include?(params[:endpoint])
+          end
 
-            if params[:endpoint] == 'launch'
-              recent_results = Inferno::Models::SequenceResult.all(
-                :created_at.gte => 5.minutes.ago,
-                :result => 'wait',
-                :order => [:created_at.desc]
-              )
-              iss_url = params[:iss]&.downcase&.split('://')&.last&.chomp('/')
-
-              matching_results = recent_results.select do |sr|
-                testing_instance_url = sr.testing_instance.url.downcase.split('://').last.chomp('/')
-                testing_instance_url == iss_url
-              end
+          def instance_id_from_cookie
+            cookies[:instance_id_test_set]&.split('/')&.first
+          end
 
-              instance = matching_results&.first&.testing_instance
-              if instance.nil?
-                instance = Inferno::Models::TestingInstance.get(cookies[:instance_id_test_set]&.split('/')&.first)
-                if instance.nil?
-                  message = "Error: No actively running launch sequences found for iss #{params[:iss]}. " \
-                            'Please ensure that the EHR launch test is actively running before attempting to launch Inferno from the EHR.'
-                  halt 500, message
-                elsif instance&.waiting_on_sequence&.wait?
-                  error_message = 'No iss for redirect'
-                else
-                  redirect "#{base_path}/#{cookies[:instance_id_test_set]}/?error=no_ehr_launch&iss=#{params[:iss]}"
-                end
-              end
+          get '/oauth2/:key/redirect/?' do
+            @instance = Inferno::Models::TestingInstance.first(state: params[:state])
+            pass if @instance.present?
```

Comment:
This `pass` should be moved to the bottom of the route implementation and happen regardless of if the instance is present.  Having it here prevents us from rescuing the test with the cookie on the next line if a state parameter isn't passed (which is the cookies only job so far).

We should pass regardless, because the next route calls `running_test_found` and handles the checking whether `@instance.present` and provides the appropriate 500 response

Reply:
I added more tests to cover these cases and adjusted the code accordingly, although not exactly how you suggested.

Reply:
Awesome, I like the way you handled it here.